File: android/src/main/java/com/reactnativemp/MpModule.java
package com.reactnativemp;

import android.util.Log;

import androidx.annotation.NonNull;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONException;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import org.json.JSONObject;

import java.util.ArrayList;

import io.dcloud.common.DHInterface.ICallBack;
import io.dcloud.feature.sdk.DCSDKInitConfig;
import io.dcloud.feature.sdk.DCUniMPCapsuleButtonStyle;
import io.dcloud.feature.sdk.DCUniMPSDK;
import io.dcloud.feature.sdk.Interface.IDCUniMPOnCapsuleMenuButtontCallBack;
import io.dcloud.feature.sdk.Interface.IDCUniMPPreInitCallback;
import io.dcloud.feature.sdk.Interface.IMenuButtonClickCallBack;
import io.dcloud.feature.sdk.Interface.IOnUniMPEventCallBack;
import io.dcloud.feature.sdk.Interface.IUniMPOnCloseCallBack;
import io.dcloud.feature.sdk.MenuActionSheetItem;
import io.dcloud.feature.unimp.DCUniMPJSCallback;

@ReactModule(name = MpModule.NAME)
public class MpModule extends ReactContextBaseJavaModule {
  public static final String NAME = "Mp";

  public MpModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  private boolean hasValue(ReadableMap rm, String key) {
    return rm.hasKey(key) && !rm.isNull(key);
  }

  private void sendEvent(String eventName, WritableMap data) {
    this.getReactApplicationContext()
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
      .emit(eventName, data);
  }

  @ReactMethod
  public void addListener(String eventName) {
    // Set up any upstream listeners or background tasks as necessary
    Log.i("addListener eventName", eventName);
  }
  @ReactMethod
  public void removeListeners(Integer count) {
    // Remove upstream listeners, stop unnecessary background tasks
    Log.i("removeListeners count", String.valueOf(count));
  }

  @ReactMethod
  public void initialize(ReadableMap params, Promise promise) {
    try {
      ReactApplicationContext cxt = this.getReactApplicationContext();
      DCSDKInitConfig.Builder builder = new DCSDKInitConfig.Builder();
      if (params != null) {
        if (hasValue(params, "menus")) {
          ReadableArray menus = params.getArray("menus");
          int len = menus.size();
          if (len > 0) {
            ArrayList<MenuActionSheetItem> sheets = new ArrayList();
            for (int i = 0; i < len; i++) {
              ReadableMap m = menus.getMap(i);
              sheets.add(new MenuActionSheetItem(m.getString("title"), m.getString("key")));
            }
            builder.setMenuActionSheetItems(sheets);
          }
        }

        if (hasValue(params, "capsule")) {
          builder.setCapsule(params.getBoolean("capsule"));
        }

        if (hasValue(params, "enableBackground")) {
          builder.setEnableBackground(params.getBoolean("enableBackground"));
        }

        if (hasValue(params, "isFromRecents")) {
          builder.setUniMPFromRecents(params.getBoolean("isFromRecents"));
        }

        if (hasValue(params, "fontSize")) {
          builder.setMenuDefFontSize(params.getString("fontSize"));
        }

        if (hasValue(params, "fontColor")) {
          builder.setMenuDefFontColor(params.getString("fontColor"));
        }

        if (hasValue(params, "fontWeight")) {
          builder.setMenuDefFontWeight(params.getString("fontWeight"));
        }

        if (hasValue(params, "capsuleButtonStyle")) {
          ReadableMap capsuleButtonStyle = params.getMap("capsuleButtonStyle");
          DCUniMPCapsuleButtonStyle style = new DCUniMPCapsuleButtonStyle();
          String backgroundColor = params.getString("backgroundColor");
          if (backgroundColor != null) {
            style.setBackgroundColor(backgroundColor);
          }
          String borderColor = params.getString("borderColor");
          if (borderColor != null) {
            style.setBorderColor(borderColor);
          }
          String highlightColor = params.getString("highlightColor");
          if (highlightColor != null) {
            style.setHighlightColor(highlightColor);
          }
          String textColor = params.getString("textColor");
          if (textColor != null) {
            style.setTextColor(textColor);
          }
          builder.setCapsuleButtonStyle(style);
        }
      }
//      DCUniMPSDK.getInstance().setCapsuleMenuButtonClickCallBack(new IDCUniMPOnCapsuleMenuButtontCallBack() {
//        @Override
//        public void menuButtonClicked(String s) {
//          cxt.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
//            .emit("onCapsuleMenuButtonClick", s);
//        }
//      });
//      DCUniMPSDK.getInstance().setOnUniMPEventCallBack(new IOnUniMPEventCallBack() {
//        @Override
//        public void onUniMPEventReceive(String s, String s1, Object o, DCUniMPJSCallback dcUniMPJSCallback) {
//          dcUniMPJSCallback.invoke(o); // 小程序js回调接口
//        }
//      });
      DCUniMPSDK.getInstance().setDefMenuButtonClickCallBack(new IMenuButtonClickCallBack() {
        @Override
        public void onClick(String appid, String key) {
          WritableMap map = Arguments.createMap();
          map.putString("key", key);
          map.putString("appid", appid);
          cxt.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit("onMenuButtonClick", map);
        }
      });
      DCUniMPSDK.getInstance().setUniMPOnCloseCallBack(new IUniMPOnCloseCallBack() {
        @Override
        public void onClose(String appid) {
          cxt.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit("onCloseMp", appid);
        }
      });

      DCUniMPSDK.getInstance().initialize(cxt, builder.build(), new IDCUniMPPreInitCallback() {
        @Override
        public void onInitFinished(boolean isSuccess) {
          promise.resolve(isSuccess);
        }
      });
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void releaseWgtToRunPathFromPath(String wgtPath, String appId, Promise promise) {
    try {
      DCUniMPSDK.getInstance().releaseWgtToRunPathFromePath(appId, wgtPath, new ICallBack() {
        @Override
        public Object onCallBack(int code, Object pArgs) {
          if (code == 1) {//释放wgt完成
            promise.resolve(pArgs);
          } else {//释放wgt失败
            promise.reject(code + pArgs.toString());
          }
          return null;
        }
      });
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void isExistsApp(String appid, Promise promise) {
    try {
      promise.resolve(DCUniMPSDK.getInstance().isExistsApp(appid));
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void launch(ReadableMap config, Promise promise) {
    try {
      String path = null;
      JSONObject params = null;
      Class splash = null;
      String appid = config.getString("appid");
      if (hasValue(config, "path")) {
        path = config.getString("path");
      }
      if (config.hasKey("path")) {
        path = config.getString("path");
      }
      if (hasValue(config, "params")) {
        params = ReactNativeJson.convertMapToJson(config.getMap("params"));
      }
      if (hasValue(config, "splash")) {
        if (config.getBoolean("splash")) {
          splash = MpSplashView.class;
        }
      }
      Log.i("launch", params.toString());
      DCUniMPSDK.getInstance().openUniMP(this.getCurrentActivity(), appid, splash, path, params);
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void getAppVersionInfo(String appid, Promise promise) {
    try {
      JSONObject info = DCUniMPSDK.getInstance().getAppVersionInfo(appid);
      if (info == null) {
        promise.resolve("");
      }
//      WritableMap map = Arguments.createMap();
//      map.putString("name", info.getString("name"));
//      map.putInt("code", new Integer(info.getString("code")));
      promise.resolve(ReactNativeJson.convertJsonToMap(info));
    } catch (Exception e) {
      promise.reject(e);
    }
  }
}

File: android/src/main/java/com/reactnativemp/MpModule.kt
package com.reactnativemp

import android.util.Log
import com.alibaba.fastjson.JSON.parseObject
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import io.dcloud.feature.sdk.DCSDKInitConfig
import io.dcloud.feature.sdk.DCUniMPSDK
import io.dcloud.feature.sdk.MenuActionSheetItem
import com.alibaba.fastjson.JSONObject
import io.dcloud.feature.sdk.Interface.IUniMP
import io.dcloud.feature.sdk.DCUniMPCapsuleButtonStyle
import io.dcloud.feature.barcode2.BarcodeProxy.context
import android.widget.Toast

class MpModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
  private var appid: String? = null
  private var path: String? = null
  private var iup: IUniMP? = null

  override fun getName(): String {
    return "Mp"
  }

  private fun sendEvent(evntName: String, params: Any) {
    var data = params
    if (params is JSONObject) {
      data = params.toString()
    }
    reactContext.getJSModule(RCTDeviceEventEmitter::class.java).emit(evntName, data)
  }

  @ReactMethod
  fun initialize(params: ReadableMap, promise: Promise) {
    try {
      val items = params.getArray("items")
      val capsule = params.getBoolean("capsule")
      val fontSize = params.getString("fontSize")
      val fontColor = params.getString("fontColor")
      val fontWeight = params.getString("fontWeight")
      val sheetItems: MutableList<MenuActionSheetItem> = ArrayList()
      for (i in 0 until items!!.size()) {
        val item = items.getMap(i)
        sheetItems.add(MenuActionSheetItem(item!!.getString("title"), item.getString("key")))
      }

      //首先构建 DCUniMPCapsuleButtonStyle胶囊按钮样式
      val style = DCUniMPCapsuleButtonStyle()

      //设置胶囊按钮背景颜色
      style.setBackgroundColor("#990099")
      //设置胶囊按钮“···｜x” 的字体颜色
      style.setTextColor("#090909")
      //设置胶囊按钮边框颜色
      style.setBorderColor("#777777")
      //设置胶囊按钮按下状态背景颜色
      style.setHighlightColor("#888888")
      //**以上目前可设置的样式**

      val config = DCSDKInitConfig.Builder().setCapsule(capsule).setMenuDefFontSize(fontSize)
        .setMenuDefFontColor(fontColor)
        .setMenuDefFontWeight(fontWeight)
        .setMenuActionSheetItems(sheetItems)
        .setCapsuleButtonStyle(style)
        .setEnableBackground(true)
        .build()
      val instance = DCUniMPSDK.getInstance()
      instance.setDefMenuButtonClickCallBack { appid, key -> sendEvent("MenuItemClick", key) }

      instance.setOnUniMPEventCallBack { appid, event, data, callback ->
        sendEvent("MPEventReceive", data)
        Log.d("cs", "onUniMPEventReceive    event=" + event);
        //回传数据给小程序
        callback.invoke("收到消息");
      }

      instance.setUniMPOnCloseCallBack { appid -> sendEvent("MPOnclose", appid) }
      instance.initialize(reactContext, config) { b -> promise.resolve(b) }
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun launch(config: ReadableMap, promise: Promise) {
    try {
      var params: org.json.JSONObject? = null
      this.appid = config.getString("appid")
      if (config.hasKey("path")) {
        this.path = config.getString("path")
      }
      if (config.hasKey("params")) {
//        params = config.getString("params")
//        params = config.getMap()

//        params = JSONObject.parseObject(config.getString("params"))
//        log(111.0, config.getString("params"))
//        val p = config.getMap("params")
//        p.
      }
//      DCUniMPSDK.getInstance().openUniMP()

//      DCUniMPSDK.getInstance().startApp(context, "__UNI__04E3A11", MpSplashView::class.java)
      iup = DCUniMPSDK.getInstance().openUniMP(this.currentActivity, this.appid, MpSplashView::class.java, this.path, params)
//      DCUniMPSDK.getInstance().startActivityForUniMPTask()
//      DCUniMPSDK.getInstance().startApp(this.currentActivity, appid, MpSplashView::class.java, path, if (params == null) null else org.json.JSONObject(params))
      promise.resolve(null)
    } catch (e: Exception) {
      e.printStackTrace()
      promise.reject(e)
    }
  }

  @ReactMethod
  fun getAppBasePath(promise: Promise) {
    try {
      val path = DCUniMPSDK.getInstance().getAppBasePath(this.currentActivity)
      promise.resolve(path)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun isExistsApp(appid: String?, promise: Promise) {
    try {
      promise.resolve(DCUniMPSDK.getInstance().isExistsApp(appid))
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun getRuningAppid(promise: Promise) {
    try {
      promise.resolve(this.appid)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun getCurrentPageUrl(promise: Promise) {
    try {
//      DCUniMPSDK.getInstance().getCurrentPageUrl()
      promise.resolve(path)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun getAppVersionInfo(appid: String?, promise: Promise) {
    try {
      val info = DCUniMPSDK.getInstance().getAppVersionInfo(appid)
      if (info === null) {
        promise.resolve(null)
        return
      }
      val map = Arguments.createMap()
      map.putString("name", info.getString("name"))
      map.putInt("code", info.getString("code").toInt())
      promise.resolve(map)
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun closeCurrentApp(promise: Promise) {
    try {
      promise.resolve(iup?.closeUniMP())
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun sendEvent(name: String?, data: ReadableMap?, promise: Promise) {
    try {
      promise.resolve(iup?.sendUniMPEvent(name, data))
    } catch (e: Exception) {
      promise.reject(e)
    }
  }

  @ReactMethod
  fun releaseWgtToRunPathFromPath(path: String?, promise: Promise) {
    try {
      val wgtPath = context.externalCacheDir!!.path + "/__UNI__04E3A11.wgt"
      DCUniMPSDK.getInstance().releaseWgtToRunPathFromePath(this.appid, path) { code, pArgs ->
        if (code == 1) { //释放wgt完成
          try {
            DCUniMPSDK.getInstance().openUniMP(context, "__UNI__04E3A11")
          } catch (e: Exception) {
            e.printStackTrace()
          }
        } else { //释放wgt失败
          Toast.makeText(context, "资源释放失败", Toast.LENGTH_SHORT).show()
        }
        null
      }

//      DCUniMPSDK.getInstance().getAppBasePath()
//      iup?.releaseWgtToRunPathFromePath(path) { code: Int, _: Any? ->
//        promise.resolve(code == 1)
//      }
    } catch (e: Exception) {
      promise.reject(e)
    }
  }
}
